# U.S.A. Presidential Vocabulary

My Codecademy portfolio project from the <a href='https://www.codecademy.com/learn/paths/data-science'>Data Scientist Path</a> Natural Languages Processing (NLP) Course, Word Embeddings Section.

## Overview

Whenever a United States of America president is elected or re-elected, an inauguration ceremony takes place to mark the beginning of the president’s term. During the ceremony, the president gives an inaugural address to the nation, dictating the tone and focus of the next four years of leadership.

In this project you will have the chance to analyze the inaugural addresses of the presidents of the United States of America, as collected by the <a href="https://www.nltk.org/book/ch02.html">Natural Language Toolkit</a>, using word embeddings.

By training sets of word embeddings on subsets of inaugural address versus the collection of presidents as a whole, we can learn about the different ways in which the presidents use language to convey their agenda.

#### Project Goal:

Analyze USA presidential inaugural speeches using NLP word embeddings models.

#### Project Requirements

Be familiar with:
- Python3
- NLP (Natural Languages Processing)
<br><br>
- The Python Libraries:
    - re
    - Pandas
    - Json
    - Collections
    - NLKT
    - gensim
  

#### Links:

<a href='https://www.alex-ricciardi.com/post/u-s-a-presidential-vocabulary'>My Project Blog Presentation<a><br>
<br>
<a href='https://github.com/ARiccGitHub/us_presidential_vocabulary'>Project GitHub<a><br>

<h2 style='color : MediumBlue'>Preprocessing the Data</h2>

The project corpus data can be freely downloaded from <a href='http://www.nltk.org/nltk_data/'>NLTK Corpora</a> under the designation "68. C-Span Inaugural Address Corpus".

#### Libraries:

In [1]:
# Regex
import re
# Operating system dependent functionality
import os
# JSON encoder and decoder
import json
# Data manipulation tool
import pandas as pd
# Natural language processing
import nltk
# Tokenization into sentences
from nltk.tokenize import PunktSentenceTokenizer
# Stop words and lexical database of English  
from nltk.corpus import stopwords, wordnet
# lemmatization class
from nltk.stem import WordNetLemmatizer
# Counter Dictionary class - https://docs.python.org/3/library/collections.html#collections.Counter -
from collections import Counter

#### Saves list function:

In this project, I want to save lists of lists, a list of lists is a list of objects, I use <a hreff='https://docs.python.org/3/library/json.html'>json</a> to save the lists in my list as objects.

The save_list() function:

- Takes the arguments:
    - file_name, string data type
    - list_to_save, list data type
<br><br>
- Saves list_to_save into file_name.txt as json objects

In [2]:
def save_list(file_name, list_to_save):   
    with open(f'data/{file_name}.txt', 'w') as file:
        file.write(json.dumps(list_to_save))

To load the list files you can use

The load_list() function:

- Takes the arguments:
    - list_name, string data type
- Load list_name.txt
<br><br>
- Returns the list_name.txt as a list 

In [3]:
def load_list(list_name):
    with open(f'data/{file_name}.txt', 'r') as file:
        return json.loads(file.read())

<h3 style='color : DarkMagenta'>Converting files into a corpus</h3>

For this project, I decided to combine the files data into a corpus that I named ```speeches```.

In [4]:
# Project directory path
path = os.getcwd()
# Sorts and save files name from the corpus_data folder
file_names = sorted([file for file in os.listdir(f"{path}/corpus_data")])
# Creates a speeches list from files   
speeches = []
for name in file_names:  
    with open(f'corpus_data/{name}', 'r+') as file:  
        speeches.append(file.read())

Sample from the ```speeches``` corpus:

In [5]:
# 1793-Washington's speech
speeches[1]

'Fellow citizens, I am again called upon by the voice of my country to execute the functions of its Chief Magistrate. When the occasion proper for it shall arrive, I shall endeavor to express the high sense I entertain of this distinguished honor, and of the confidence which has been reposed in me by the people of united America.\n\nPrevious to the execution of any official act of the President the Constitution requires an oath of office. This oath I am now about to take, and in your presence: That if it shall be found during my administration of the Government I have in any instance violated willingly or knowingly the injunctions thereof, I may (besides incurring constitutional punishment) be subject to the upbraidings of all who are now witnesses of the present solemn ceremony.\n\n \n'

We can see from the ```speeches``` corpus sample that the speech's texts is not clean, it can not be processed properly by a NLT model.

<h3 style='color : DarkMagenta'>Preprocessing corpus</h3>

Before a text can be processed by a NLP model, the text data needs to be preprocessed.<br>
Text data preprocessing is the process of cleaning and prepping  the text data to be processed by NLP models.

Cleaning and prepping tasks:

- Noise removal is a text pre-processing step concerned with removing unnecessary formatting from our text.

- Tokenization is a text pre-processing step devoted to breaking up text into smaller units (usually words or discrete terms).
<br>

- Normalization is the name we give most other text preprocessing tasks, including stemming, lemmatization, upper and lowercasing, and stopword removal.

    - Stemming is the normalization preprocessing task focused on removing word affixes. 

	- Lemmatization is the normalization preprocessing task that more carefully brings words down to their root forms.

#### Tokenization

In this project, I break down the presidents' speeches into words on a sentence by sentence basis by using the sentence tokenizer <a href='https://www.nltk.org/_modules/nltk/tokenize/punkt.html'>nltk.tokenize.PunktSentenceTokenizer()</a> class.

#### Part-of-Speech Tagging

To improve the performance of <a href="https://nlp.stanford.edu/IR-book/html/htmledition/stemming-and-lemmatization-1.html">lemmatization</a> (bring a word to his root), each word in the processed text is assigned parts of speech tag, 
<a href="https://nlp.stanford.edu/software/tagger.shtml#:~:text=A%20Part%2DOf%2DSpeech%20Tagger,like%20'noun%2Dplural'.">Part-of-Speech Tagging</a> is the process of reading text in some language and assigns parts of speech to each word (and other token), such as noun, verb, adjective, etc.

Part-of-Speech tagging function:

The ```get_part_of_speech()``` function:
- Takes the arguments:
    - ```word```, string data type.<br>
<br>
- Matches ```word``` with synonyms
- Tags ```word``` and count tags.<br> 
<br>
- Returns The most common tag, the tag with the highest count, ex: n for Noun, string data type.

In [6]:
def get_part_of_speech(word):
    # Synonyms matching
    probable_part_of_speech = wordnet.synsets(word)
    # Initializing Counter class object
    pos_counts = Counter()
    # Taging and counting tags
    pos_counts["n"] = len(  [ item for item in probable_part_of_speech if item.pos()=="n"]  ) # Noun
    pos_counts["v"] = len(  [ item for item in probable_part_of_speech if item.pos()=="v"]  ) # Verb
    pos_counts["a"] = len(  [ item for item in probable_part_of_speech if item.pos()=="a"]  ) # Adjectif
    pos_counts["r"] = len(  [ item for item in probable_part_of_speech if item.pos()=="r"]  ) # Adverb
    # The most common tag, the tag with the highest count, ex: n for Noun 
    most_likely_part_of_speech = pos_counts.most_common(1)[0][0]
    
    return most_likely_part_of_speech

#### The word 'us':

The word 'us' is a commonly used word in presidential inauguration addresses.<br>
The word 'us' is a commonly used word in presidential inauguration addresses.
The result of preprocessing the word 'us' through lemmatizing with the part-of-speech tagging method <a href='https://www.nltk.org/api/nltk.corpus.reader.html#nltk.corpus.reader.wordnet.Lemma.synset'>nlt.corpus.reader.wordnet.synsets()</a> function and in conjunction with `stopwords` removal and the <a href='https://www.nltk.org/_modules/nltk/stem/wordnet.html'>nltk.stem.WordNetLemmatizer().lemmatize()</a> method, is that the word 'us' becomes 'u'.<br>
<br>
This happens because the lemmatize(word, get_part_of_speech(word)) method removes the character 's' at the end of words tagged as nouns. The word 'us', which is not part of the stopwords list, is tagged as a noun causing the lemmatization result of 'us' to be 'u'.

The word `'us'` is not a `stopword`:

In [7]:
'us' in set(stopwords.words('english'))

False

The `get_part_of_speech()` function tags the word `'us'` as a noun:

In [8]:
get_part_of_speech('us')

'n'

The `lemmatize(word, get_part_of_speech(word))` method removes the character `'s'` at the of `words` tags as noun, and the word `'us'` is tagged as a noun causing the lemmatization result of `'us'` to be `'u'`.   

In [9]:
normalizer_us = WordNetLemmatizer()
normalizer_us.lemmatize('us', get_part_of_speech('us'))

'u'

In [10]:
print(f'{normalizer_us.lemmatize("us", get_part_of_speech("us"))}\n')

u



#### Preprocessing:

In [11]:
# Stop words
stop_words = set(stopwords.words('english'))
# Initializes the lemmatizer
normalizer = WordNetLemmatizer()
# Creates an empty list of processed speeches
preprocessed_speeches = []
# ---------------------------------------------------- Preprocessing loop
for speech in speeches:
    # ------------------ Tokenizing
    # Initializes sentence tokenizer
    sentence_tokenizer = PunktSentenceTokenizer()
    # Tokenizes speech into sentences
    sentence_tokenized_speech = sentence_tokenizer.tokenize(speech)
    # ------------------ Normalizing loop
    # Creates an empty sentences list 
    word_sentences = [] 
    for sentence in sentence_tokenized_speech:
        # ----------- Removes noise from sentence and tokenizes the sentence into words  
        word_tokenized_sentence = [re.sub('[^a-zA-Z0-9]+', '', word.lower()) \
                                           for word in sentence.replace(",","").replace("-"," ").replace(":","").split()] 
        # ---------------- Removes stopwords from sentences
        sentence_no_stopwords = [word for word in word_tokenized_sentence if word not in stop_words]
        # ---------------- Before lemmatizing, adds a 's' to the word 'us'
        word_sentence_us = ['uss' if word == 'us' else word for word in sentence_no_stopwords]
        # ---------------- Lemmatizes
        word_sentence = [normalizer.lemmatize(word, get_part_of_speech(word)) \
                                                        for word in word_sentence_us if not re.match(r'\d+', word)]
        # Stores preprocessed word   
        word_sentences.append(word_sentence)     
    # Stores sentence tokenized into words     
    preprocessed_speeches.append(word_sentences) 
# Saves preprocessed corpus
save_list('preprocessed_speeches', preprocessed_speeches)

A personal preference <a href='https://docs.python.org/3/library/pprint.html'>pprint-Data pretty printer</a>

In [12]:
%pprint

Pretty printing has been turned OFF


Sample from the ```preprocessed_speeches``` list, preprocessed corpus:

In [13]:
# Dispays words from the second speech first sentences
preprocessed_speeches[1][0]

['fellow', 'citizen', 'call', 'upon', 'voice', 'country', 'execute', 'function', 'chief', 'magistrate']

Dictionary of the presidents' words speeches by sentences:

In [14]:
# Creates a list of the speech's names relative to the presidents' names and year of the speech 
year_president_speech_names = [name.lower().replace('.txt', '').replace('1989-bush', '1989-bush senior') for name in file_names]
# Creates a dictionary of the presidents preprocessed speeches
presidents_speeches = dict(zip(year_president_speech_names, preprocessed_speeches))

Presidents pre-processed speeches DataFrame:

In [15]:
df_presidents_speeches = pd.DataFrame({'Preprocessed Speech' : preprocessed_speeches}, index = year_president_speech_names)
df_presidents_speeches.to_csv('data/processed_presidents_speeches.csv')
df_presidents_speeches.head()

Preprocessed Speech
1789-washington  [[fellow, citizen, senate, house, representati...
1793-washington  [[fellow, citizen, call, upon, voice, country,...
1797-adams       [[first, perceive, early, time, middle, course...
1801-jefferson   [[friend, fellow, citizen, call, upon, underta...
1805-jefferson   [[proceed, fellow, citizen, qualification, con...

Sample from the ```df_presidents_speeches``` DataFrame:

In [16]:
# Dispays words in each sentences from the 1793-Washington's speech 
df_presidents_speeches.loc['1793-washington'][0]

[['fellow', 'citizen', 'call', 'upon', 'voice', 'country', 'execute', 'function', 'chief', 'magistrate'], ['occasion', 'proper', 'shall', 'arrive', 'shall', 'endeavor', 'express', 'high', 'sense', 'entertain', 'distinguish', 'honor', 'confidence', 'repose', 'people', 'unite', 'america'], ['previous', 'execution', 'official', 'act', 'president', 'constitution', 'require', 'oath', 'office'], ['oath', 'take', 'presence', 'shall', 'find', 'administration', 'government', 'instance', 'violate', 'willingly', 'knowingly', 'injunction', 'thereof', 'may', 'besides', 'incur', 'constitutional', 'punishment', 'subject', 'upbraiding', 'witness', 'present', 'solemn', 'ceremony']]

In [17]:
# Dispays words in the first sentence from the 1793-Washington's speech 
df_presidents_speeches.loc['1793-washington'][0][0]

['fellow', 'citizen', 'call', 'upon', 'voice', 'country', 'execute', 'function', 'chief', 'magistrate']

Combine list of all the sentences from all the president speeches  

In [18]:
# Creates an empty list of all the stences in processed_speeches
all_sentences = [sentence for speech in preprocessed_speeches for sentence in speech]
# Saves all_sentences
save_list('all_sentences', all_sentences)

Sample from the ```all_sentences``` list:

In [19]:
all_sentences[23]

['fellow', 'citizen', 'call', 'upon', 'voice', 'country', 'execute', 'function', 'chief', 'magistrate']

Words in all sentences list:

In [20]:
all_words = [word for sentence in all_sentences for word in sentence]
# Saves all_words
save_list('all_words', all_words)

<h2 style='color : MediumBlue'>Word Embeddings</h2>

<a href='https://www.codecademy.com/learn/natural-language-processing/modules/nlp-word-embeddings'>Word embeddings</a> are a type of word representation that allows words with similar meaning to have a similar representation. In NLP words are often represented as numeric vectors, the algorithms used to vectorize words are referred to as "words to vectors"(<a href='https://en.wikipedia.org/wiki/Word2vec'>word2vec</a>).

#### Libraries:
In addition of the libraries imported for pre-processing the data, I use the fallowing libraries for Word Embeddings

In [21]:
# word2vec model library  
import gensim

Note: 
<a href='https://blog.usejournal.com/how-does-the-model-produce-different-results-on-same-dataset-54486f951dbf'>Machine learning models will produce different results on same dataset</a>, the models generate a sequence of random numbers called <a href='https://towardsdatascience.com/how-to-use-random-seeds-effectively-54a4cd855a79'>random seed</a> used within the process of generating test, validation and training datasets from a given dataset. Configurating a model's seed to a set value will ensure that the results are reproducible.<br>
The  python library `gensim` relies on different processes to initialize and train its word embeddings model class, if you need to generate more consistent results (not recommended), click on the following link:<br>
<a href='https://stackoverflow.com/questions/34831551/ensure-the-gensim-generate-the-same-word2vec-model-for-different-runs-on-the-sam'>Ensure the gensim generate the same Word2Vec model for different runs on the same data</a>

#### Input Function:

The input function is optional, a personal preference, I created the function to easily input different variable values without having to change the code.<br>
The option to use the function is by default turn off. 

In [22]:
input_option = False

The input_word() function:

- Takes the arguments:
    - input_subject, string data type
    - word_list, list data integer type
<br><br>
- Outputs on screen the input_subject
- Take a user input, inputted_word
- Compares inputted_word with items in the word_list
<br><br>
- Returns inputted_word.lower()

In [23]:
 def input_word(input_subject, word_list):
        # User Input a word
        inputted_word = input(f'\nEnter a {input_subject}: ')

        while inputted_word.lower() not in word_list:
            print(f'\n{inputted_word} is not in the {input_subject} list')
            inputted_word = input(f'\nPlease reenter a {input_subject}: ')    
        print()

        return inputted_word.lower()

<h3 style='color : DarkMagenta'>All Presidents</h3>

Analysis of the presidential vocabulary by looking at all the inaugural addresses.

Most frequently used terms:<br>
The following list of words are the ten most frequently used presidential inauguration speech terms.  
The numeric values represent the number of times the corresponding words appeared in the combined  presidential inauguration speeches.  

In [24]:
most_freq_words = Counter(all_words).most_common()
# Saves most_freq_words
save_list('most_freq_words', most_freq_words)

# 10 most frequently used words
most_freq_words[:10]

[('government', 651), ('people', 623), ('nation', 515), ('us', 480), ('state', 448), ('great', 394), ('upon', 371), ('must', 366), ('make', 357), ('country', 355)]

In [25]:
# 3 most frequently used words with count
most_freq_words[:3]

[('government', 651), ('people', 623), ('nation', 515)]

The fallowing list of words are the three most frequently used presidential inauguration speeches' terms.  

In [26]:
# 3 most frequently used words
[word[0] for word in most_freq_words[:3]]

['government', 'people', 'nation']

From the three most frequently used term results, we can see that the main topic within the combined inaugural addresses seem to be centered around the terms `government`, `people` and `nation`.<br>

#### Word2Vec, word embeddings model

The idea behind word embeddings is a theory known as the distributional hypothesis. This hypothesis states that words that co-occur in the same contexts tend to have similar meanings.
Word2Vec is a shallow neural network model that can build word embeddings using either continuous bag-of-words or continuous skip-grams.<br>
<br>
The word2vec method that I use to create word embeddings is based on continuous skip-grams. Skip-grams function similarly to n-grams, except instead of looking at groupings of n-consecutive words in a text, we can look at sequences of words that are separated by some specified distance between them.<br>
<br>
For this project, we want to create word embeddings model using the skip-grams word2vec model, within the USA presidential inaugural speeches context.    

In [27]:
word_embeddings = gensim.models.Word2Vec(all_sentences, size=96, window=5, min_count=1, workers=2, sg=1)

Note: ```gensim.models.Word2Vec()``` takes a text as an argument to give context to the words, for this project the sentences in ```all_sentences``` list are the context.

Vocabulary of terms:<br>
For this project, the vocabulary of terms is the list unique words from the words in all sentences list, ```all_words```.<br>
In other words, the vocabulary of terms is the list of words, which are not stop words, use within the inaugural speeches. 

In [28]:
# Removing duplicated words in all_words
vocabulary_of_terms = list(set(all_words))
# Saves vocabulary_of_terms
save_list('vocabulary_of_terms', vocabulary_of_terms)

A sample of a word vector representation generate by the ```word_embedding``` model.  

In [29]:
vec_word = 'us'

# Word verctor representation
print(vec_word)
print(word_embeddings.wv[vec_word])

us
[-0.03566258 -0.057491   -0.22429784 -0.19385831 -0.05224766 -0.21745294
 -0.03083961  0.01476462 -0.35729468  0.07782798  0.15073359 -0.01993579
 -0.16864842  0.12328792 -0.22607584 -0.02692644 -0.17266034 -0.2122943
  0.0932364  -0.20162335 -0.08800333  0.13730125 -0.27899835 -0.14394923
  0.07692517  0.13672738  0.03948302  0.27413663 -0.3095462  -0.05113837
 -0.05605842  0.09483039 -0.15032959 -0.05592294  0.00599146 -0.2568221
  0.1071072   0.11913454  0.11639356 -0.13132544 -0.04766015 -0.11872021
  0.28670758 -0.26508254 -0.01220568  0.16427317  0.2563698  -0.17130415
 -0.11297461 -0.12608878 -0.01350829 -0.24418342 -0.04711331  0.20453948
  0.10789154 -0.28325412  0.00551981  0.05497228  0.20139584 -0.06281348
  0.10151973  0.2364551  -0.33533013  0.08800003 -0.02218356 -0.12237483
 -0.38471484  0.03775231  0.12288336  0.20087232  0.26013163 -0.03415838
  0.16984472 -0.16185957 -0.14474404  0.10821487 -0.07793511  0.09060979
  0.35984805 -0.22210045 -0.23348917 -0.07200254  

Using the word vectors we can calculate the cosine distances between the vectors to find out how similar terms are within the USA presidential inaugural speeches context.

Similar terms sample:<br>
Using the word vectors created by word embeddings model, we can calculate the cosine distances between the vectors to find out how similar terms are within the USA presidential inaugural speeches context.

In [30]:
# Optional input function
if input_option:
    similar_to_word = input_word('similar word', vocabulary_of_terms)
else:
    similar_to_word = 'government'
    
# Similar to 
print(similar_to_word)
# Calculate the cosine distance between word vectors outputting the 20 most similar words to the inputted word
similar_word_dist_vec = word_embeddings.wv.most_similar(similar_to_word, topn=20)
# Saves vocabulary_of_terms
save_list('vocabulary_of_terms', vocabulary_of_terms)
# List of similar words and their vectors cosine distance relative to the inputted word
print(similar_word_dist_vec) 

government
[('power', 0.9972726702690125), ('federal', 0.9970261454582214), ('authority', 0.9962713122367859), ('support', 0.9958561062812805), ('executive', 0.9952426552772522), ('grant', 0.9951847791671753), ('within', 0.9951294660568237), ('exercise', 0.9944376349449158), ('law', 0.9943814277648926), ('territory', 0.9943108558654785), ('protect', 0.9942190051078796), ('general', 0.9940023422241211), ('defend', 0.9939836859703064), ('respect', 0.9939656853675842), ('reserve', 0.9939048290252686), ('union', 0.9937343001365662), ('preserve', 0.993715763092041), ('principle', 0.9934515357017517), ('limit', 0.9931932091712952), ('local', 0.9931504726409912)]


In [31]:
# List of the similar words no cosine distance
print(similar_to_word)
print([word[0] for word in similar_word_dist_vec])

government
['power', 'federal', 'authority', 'support', 'executive', 'grant', 'within', 'exercise', 'law', 'territory', 'protect', 'general', 'defend', 'respect', 'reserve', 'union', 'preserve', 'principle', 'limit', 'local']


From  the presidential inaugural addresses data and  by training a word embeddings model with it, I was able to create a some what accurate U.S.A. presidential vocabulary,  the small size of corpus limits how efficiently the model can be trained, nonetheless the model gives us good insight into how terms are connected to each other within the presidential inauguration addresses context. 

<h3 style='color : DarkMagenta'>One President</h3>

Analysis of a president vocabulary by looking at his inaugural addresses.

Preprocessing president names:<br>
From the list `year_president_speech_names`, I can extract the president names, but the list as duplicated president names, ex: `['1789-washington', '1792-washington']`<br>
After removing the the years from the `year_president_speech_names` values, I could used `set()` to remove duplicated values, but `set(`) does not preserve the list values insertion order and I want to keep the values insertion order as `['washington', ..., ..., ... , 'trump']`, the best method to remove a list duplicated values and preserve the values insertion order is to use a `dictionary` as follow:

In [32]:
president_names = list(dict.fromkeys([re.sub(r'^....-', '', name) for name in year_president_speech_names]))
# Saves president_names
save_list('president_names', president_names)

president_names

['washington', 'adams', 'jefferson', 'madison', 'monroe', 'jackson', 'vanburen', 'harrison', 'polk', 'taylor', 'pierce', 'buchanan', 'lincoln', 'grant', 'hayes', 'garfield', 'cleveland', 'mckinley', 'roosevelt', 'taft', 'wilson', 'harding', 'coolidge', 'hoover', 'truman', 'eisenhower', 'kennedy', 'johnson', 'nixon', 'carter', 'reagan', 'bush senior', 'clinton', 'bush', 'obama', 'trump']

In [33]:
# Optional input function
if input_option:
    president_name = input_word('president name', president_names)
else:
    president_name = 'madison'


# Speeches list
one_president_speeches = [presidents_speeches[name] for name in year_president_speech_names if president_name in name]
# Sentences list 
one_president_sentences = [sentence for speech in one_president_speeches for sentence in speech]
# Words list
one_president_all_words = [word for sentence in one_president_sentences for word in sentence]

The president most frequently used terms:

In [34]:
one_president_most_freq_words = Counter(one_president_all_words).most_common()
# 10 most frequently used words
print(president_name)
print(one_president_most_freq_words[:10])

madison
[('war', 17), ('nation', 13), ('country', 11), ('state', 10), ('public', 8), ('unite', 8), ('right', 7), ('every', 7), ('without', 6), ('long', 6)]


In [35]:
# 3 most frequently used words
print(president_name)
print([word[0] for word in one_president_most_freq_words[0:3]])

madison
['war', 'nation', 'country']


The president word embeddings model

In [36]:
one_president_word_embeddings = gensim.models.Word2Vec(one_president_sentences, size=96, window=5, min_count=1, workers=2, sg=1)

The president vocabulary of terms:

In [37]:
# Removing duplicated words in one_president_all_words
one_president_vocabulary_of_terms = list(set(one_president_all_words))

How similar terms are within the president's presidential inaugural speeches context.

In [38]:
# Optional input function
if input_option:
    one_president_similar_to_word = input_word('word', one_president_vocabulary_of_terms)
else:
    one_president_similar_to_word = 'government'
    
# Similar to 
print(f'{president_name}\'s {one_president_similar_to_word}')
# Calculate the cosine distance between word vectors outputting the 20 most similar words to the inputted word
one_president_similar_word_dist = one_president_word_embeddings.wv.most_similar(one_president_similar_to_word, topn=20)
# List of similar words and their vectors cosine distance relative to the inputted word
print(one_president_similar_word_dist)

madison's government
[('discontinue', 0.3531048893928528), ('weight', 0.31632769107818604), ('strengthen', 0.3040248155593872), ('perpetuate', 0.292847603559494), ('encourage', 0.28303566575050354), ('emigrate', 0.26809194684028625), ('knife', 0.2605974078178406), ('noble', 0.2598712742328644), ('emanate', 0.2519160807132721), ('mariner', 0.24878022074699402), ('determine', 0.24856646358966827), ('native', 0.23511092364788055), ('protection', 0.23499499261379242), ('american', 0.23039644956588745), ('stamp', 0.23036402463912964), ('form', 0.22654259204864502), ('intrigue', 0.21648332476615906), ('author', 0.21170350909233093), ('revenue', 0.21013860404491425), ('support', 0.2087252289056778)]


The cosine distance values of the most similar words are under 0.5, the result are less than satisfying due to the small size of the corpus, ```one_president_sentences```, used to train the word embeddings model.

The list of the similar terms with no cosine distance:

In [39]:
print(f'{president_name}\'s {one_president_similar_to_word}')
print([word[0] for word in one_president_similar_word_dist])

madison's government
['discontinue', 'weight', 'strengthen', 'perpetuate', 'encourage', 'emigrate', 'knife', 'noble', 'emanate', 'mariner', 'determine', 'native', 'protection', 'american', 'stamp', 'form', 'intrigue', 'author', 'revenue', 'support']


I thought that it will good idea to create a presidents' vocabularies DataFrame :

In [40]:
# Creates a DataFrame 
df_presidents_vocabularies = pd.DataFrame(index=president_names)

# Sepeeches list
all_presidents_speeches = [[presidents_speeches[name] for name in year_president_speech_names if president in name] \
                                                                                        for president in president_names]
# Sentences list 
all_presidents_sentences = [[sentence for speech in speeches for sentence in speech] \
                                                                        for speeches in all_presidents_speeches]
# Words list
all_presidents_all_words = [[word for sentence in sentences for word in sentence] \
                                                                        for sentences in all_presidents_sentences]

# Each president most three recurrent words 
df_presidents_vocabularies['Three Most  Recurrent Terms'] = [[word[0] for word in Counter(words).most_common()[:3]] \
                                                                                    for words in all_presidents_all_words]
 
# Each president most 10 recurrent words 
df_presidents_vocabularies['Ten Most Recurrent Terms'] = [[word[0] for word in Counter(words).most_common()[:15]] \
                                                                                      for words in all_presidents_all_words]
# Each president vocabulary of terms
df_presidents_vocabularies['Terms List'] = [list(set(one_president_all_words)) \
                                                                   for presidents_all_words in all_presidents_all_words]
# Saves DataFrame
df_presidents_vocabularies.to_csv('data/presidents_vocabularies.csv')
df_presidents_vocabularies

Three Most  Recurrent Terms  \
washington             [government, every, may]   
adams              [government, nation, people]   
jefferson                [may, public, citizen]   
madison                  [war, nation, country]   
monroe               [state, great, government]   
jackson             [government, people, state]   
vanburen               [people, every, country]   
harrison                 [power, people, state]   
polk                 [government, state, union]   
taylor                [shall, government, duty]   
pierce                     [upon, right, power]   
buchanan           [state, shall, constitution]   
lincoln            [state, constitution, union]   
grant                   [country, best, nation]   
hayes               [country, government, upon]   
garfield     [government, people, constitution]   
cleveland          [people, government, public]   
mckinley             [upon, government, people]   
roosevelt                  [nation, people, us]   
taft               [government, make, business]   
wilson                      [upon, life, great]   
harding                     [world, must, make]   
coolidge                 [country, great, must]   
hoover                [government, law, people]   
truman                  [nation, world, people]   
eisenhower              [people, world, nation]   
kennedy                        [let, us, world]   
johnson                    [nation, us, change]   
nixon                          [us, world, let]   
carter                      [nation, new, must]   
reagan                 [us, government, people]   
bush senior                     [new, us, make]   
clinton                        [us, new, world]   
bush                 [america, freedom, nation]   
obama                        [us, must, nation]   
trump              [america, american, country]   

                                      Ten Most Recurrent Terms  \
washington   [government, every, may, citizen, present, cou...   
adams        [government, nation, people, union, upon, coun...   
jefferson    [may, public, citizen, us, government, fellow,...   
madison      [war, nation, country, state, public, unite, r...   
monroe       [state, great, government, war, citizen, unite...   
jackson      [government, people, state, power, public, uni...   
vanburen     [people, every, country, institution, governme...   
harrison     [power, people, state, government, upon, const...   
polk         [government, state, union, power, would, one, ...   
taylor       [shall, government, duty, interest, country, h...   
pierce       [upon, right, power, nation, state, government...   
buchanan     [state, shall, constitution, may, government, ...   
lincoln      [state, constitution, union, law, shall, gover...   
grant        [country, best, nation, office, people, questi...   
hayes        [country, government, upon, party, state, publ...   
garfield     [government, people, constitution, make, state...   
cleveland    [people, government, public, citizen, us, shal...   
mckinley     [upon, government, people, must, congress, gre...   
roosevelt    [nation, people, us, government, life, must, m...   
taft         [government, make, business, law, must, may, s...   
wilson       [upon, life, great, nation, shall, thing, men,...   
harding      [world, must, make, america, war, government, ...   
coolidge     [country, great, must, nation, government, peo...   
hoover       [government, law, people, nation, upon, progre...   
truman       [nation, world, people, peace, freedom, free, ...   
eisenhower   [people, world, nation, free, peace, freedom, ...   
kennedy      [let, us, world, side, power, new, nation, ple...   
johnson      [nation, us, change, must, man, people, union,...   
nixon        [us, world, let, peace, america, new, nation, ...   
carter       [nation, new, must, us, strength, people, toge...   
reagan       [us, government, people, world, american, one,...   
bush senior  [new, u

<h3 style='color : DarkMagenta'>Selection of Presidents</h3>

We can analyze further, using word embeddings, the presidential vocabulary by combining the first five US presidents' inaugural speeches and compare the results with the results from last five US presidents' inaugural speeches.

Preprocessing the data:

In [41]:
# All words list
first_5_presidents_all_words = [word for words in all_presidents_all_words[:5] for word in words]
last_5_presidents_all_words = [word for words in all_presidents_all_words[len(all_presidents_all_words)-6:-1] \
                                                                                                       for word in words]

# Sentences list
first_5_presidents_sentences = [sentence for sentences in all_presidents_sentences[:5] for sentence in sentences]
last_5_presidents_sentences = [sentence for sentences in all_presidents_sentences[len(all_presidents_sentences)-6:-1] \
                                                                                                for sentence in sentences]
# Vocabulary of terms:
first_5_presidents_vocabulary = list(set(first_5_presidents_all_words))
last_5_presidents_vocabulary = list(set(last_5_presidents_all_words))

Most frequently used terms:

In [42]:
# First five presidents
first_5_presidents_most_freq_words = Counter(first_5_presidents_all_words).most_common()
# 10 most frequently used words
print('First Five Presidents')
print(first_5_presidents_most_freq_words[:10])

# Last five presidents
last_5_presidents_most_freq_words = Counter(last_5_presidents_all_words).most_common()
# 10 most frequently used words
print('\nLast Five Presidents')
print(last_5_presidents_most_freq_words[:10])

First Five Presidents
[('government', 105), ('state', 103), ('nation', 81), ('great', 75), ('may', 69), ('citizen', 66), ('country', 65), ('people', 64), ('war', 61), ('public', 60)]

Last Five Presidents
[('us', 176), ('america', 111), ('must', 105), ('nation', 104), ('world', 101), ('new', 101), ('american', 95), ('time', 91), ('people', 90), ('freedom', 81)]


In [43]:
# First five presidents
first_5_presidents_most_freq_words = Counter(first_5_presidents_all_words).most_common()
# 3 most frequently used words
print('First Five Presidents')
print(first_5_presidents_most_freq_words[:3])

# Last five presidents
last_5_presidents_most_freq_words = Counter(last_5_presidents_all_words).most_common()
# 3 most frequently used words
print('\nLast Five Presidents')
print(last_5_presidents_most_freq_words[:3])

First Five Presidents
[('government', 105), ('state', 103), ('nation', 81)]

Last Five Presidents
[('us', 176), ('america', 111), ('must', 105)]


In [44]:
# 3 first most frequently used words

print('First Five Presidents')
print([word[0] for word in first_5_presidents_most_freq_words[:3]])

print('\nLast Five Presidents')
print([word[0] for word in last_5_presidents_most_freq_words[:3]])

First Five Presidents
['government', 'state', 'nation']

Last Five Presidents
['us', 'america', 'must']


Word embeddings:

In [45]:
first_5_presidents_word_embeddings = gensim.models.Word2Vec(first_5_presidents_sentences, size=96, window=5, 
                                                                                                   min_count=1, workers=2, sg=1)
last_5_presidents_word_embeddings = gensim.models.Word2Vec(last_5_presidents_sentences, size=96, window=5, 
                                                                                                   min_count=1, workers=2, sg=1)

Similar words:

In [46]:
# Optional input function
input_option = False
if input_option:
    first_last_pre_voc = list(set(first_5_presidents_vocabulary + last_5_presidents_vocabulary))
    first_last_pre_similar_to_word = input_word('First and last four presidents word', first_last_pre_voc)
else:
    first_last_pre_similar_to_word = 'government'

# Calculate the cosine distance between word vectors outputting the 20 most similar words to the inputted word
first_5_pre_similar_word_dist = first_5_presidents_word_embeddings.wv.most_similar(first_last_pre_similar_to_word, topn=20)
last_5_pre_similar_word_dist = last_5_presidents_word_embeddings.wv.most_similar(first_last_pre_similar_to_word, topn=20)

# List of similar words and their vectors cosine distance relative to the inputted word
print(f'First five presidents {first_last_pre_similar_to_word}')
print(first_5_pre_similar_word_dist)
print(f'\nLast five presidents {first_last_pre_similar_to_word}')
print(last_5_pre_similar_word_dist)

First five presidents government
[('great', 0.9993283152580261), ('war', 0.9993203282356262), ('may', 0.9992838501930237), ('make', 0.9992183446884155), ('state', 0.9992141723632812), ('every', 0.9992128610610962), ('nation', 0.9992037415504456), ('union', 0.9991971254348755), ('would', 0.99918532371521), ('power', 0.9991679191589355), ('public', 0.9991235136985779), ('us', 0.999123215675354), ('duty', 0.9991223812103271), ('year', 0.9991199970245361), ('country', 0.999106228351593), ('citizen', 0.9990841746330261), ('well', 0.9990724325180054), ('interest', 0.9990679621696472), ('principle', 0.9990620017051697), ('shall', 0.9990471005439758)]

Last five presidents government
[('american', 0.9995585680007935), ('world', 0.9994905591011047), ('us', 0.999477207660675), ('make', 0.9994717836380005), ('citizen', 0.9994479417800903), ('long', 0.9994418025016785), ('must', 0.999437153339386), ('great', 0.9994307160377502), ('nation', 0.9994282126426697), ('work', 0.9994194507598877), ('ameri

The cosine distance values of the most similar words are close to 1, the results are satisfying, better than ones from one president, the results are better due to the larger size of the corpus used to train the word embeddings models.

Most similar words, no cosine distances:

In [47]:
print(f'First five presidents {first_last_pre_similar_to_word}')
print([word[0] for word in first_5_pre_similar_word_dist])
print(f'\nLast five presidents {first_last_pre_similar_to_word}')
print([word[0] for word in last_5_pre_similar_word_dist])

First five presidents government
['great', 'war', 'may', 'make', 'state', 'every', 'nation', 'union', 'would', 'power', 'public', 'us', 'duty', 'year', 'country', 'citizen', 'well', 'interest', 'principle', 'shall']

Last five presidents government
['american', 'world', 'us', 'make', 'citizen', 'long', 'must', 'great', 'nation', 'work', 'america', 'life', 'freedom', 'every', 'see', 'time', 'word', 'history', 'power', 'generation']
